### PR TITLE
Check BLS changes when requesting from pool

### PR DESCRIPTION
**Pull request by potuz**
This PR checks BLS changes when requesting them from the pool (this is performed when proposing a block). |

It checks the validity of the operation individually (except signatures)
If any is found invalid it's removed from the pool
Then it checks the signatures in batch
If the batch is found invalid it checks signatures individually
If any signature is found invalid it removes it from the pool.

If a change is invalid then we continue scanning until filling the block or exhausting the pool. If a signature is invalid we do not rescan.

### CODE REVIEW

File: beacon-chain/operations/blstoexec/pool.go [hunk 1]
```diff
@@ -4,17 +4,22 @@ import (
...
 	doublylinkedlist "github.com/prysmaticlabs/prysm/v3/container/doubly-linked-list"
+	"github.com/prysmaticlabs/prysm/v3/crypto/bls/blst"
 	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+	"github.com/sirupsen/logrus"
 )
 
 // PoolManager maintains pending and seen BLS-to-execution-change objects.
 // This pool is used by proposers to insert BLS-to-execution-change objects into new blocks.
 type PoolManager interface {
 	PendingBLSToExecChanges() ([]*ethpb.SignedBLSToExecutionChange, error)
```
Comment (terencechain) [inline beacon-chain/operations/blstoexec/pool.go:21]: Shouldn't you apply this for `PendingBLSToExecChanges` or we just simply care less for that?

File: beacon-chain/operations/blstoexec/pool.go [hunk 2]
```diff
@@ -58,25 +63,69 @@ func (p *Pool) PendingBLSToExecChanges() ([]*ethpb.SignedBLSToExecutionChange, e
...
+		change, err := node.Value()
 		if err != nil {
+			p.lock.RUnlock()
 			return nil, err
 		}
+		_, err = blocks.ValidateBLSToExecutionChange(st, change)
+		if err != nil {
+			logrus.Warning("removing invalid BLSToExecutionChange from pool")
+			// MarkIncluded removes the invalid change from the pool
+			p.lock.RUnlock()
```
Comment (terencechain) [inline beacon-chain/operations/blstoexec/pool.go:81]: These lock/unlock scares me. Any reason to pick this over `defer`?

File: beacon-chain/operations/blstoexec/pool.go [hunk 3]
```diff
@@ -58,25 +63,69 @@ func (p *Pool) PendingBLSToExecChanges() ([]*ethpb.SignedBLSToExecutionChange, e
...
-		result[i], err = node.Value()
+	for node != nil && len(result) < length {
+		change, err := node.Value()
 		if err != nil {
+			p.lock.RUnlock()
 			return nil, err
 		}
+		_, err = blocks.ValidateBLSToExecutionChange(st, change)
+		if err != nil {
+			logrus.Warning("removing invalid BLSToExecutionChange from pool")
```
Comment (terencechain) [inline beacon-chain/operations/blstoexec/pool.go:79]: Should we include the err in this warning log?

File: beacon-chain/operations/blstoexec/pool.go [hunk 4]
```diff
@@ -58,25 +63,69 @@ func (p *Pool) PendingBLSToExecChanges() ([]*ethpb.SignedBLSToExecutionChange, e
...
 	}
-	return result, nil
+	p.lock.RUnlock()
+	if len(result) == 0 {
+		return result, nil
+	}
+	// We now verify the signatures in batches
+	cSet, err := blocks.BLSChangesSignatureBatch(st, result)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get BLSToExecutionChanges signatures")
```
Comment (terencechain) [inline beacon-chain/operations/blstoexec/pool.go:102]: Optional: I think you can continue here and check signatures individually

**Review by terencechain — CHANGES_REQUESTED** [review]
I have some questions

File: beacon-chain/operations/blstoexec/pool.go (hunk 1 above)
Comment (potuz) [reply beacon-chain/operations/blstoexec/pool.go:21]: I can see the value in returning the full list, someone may be interested in knowing what are all changes in the node. No one uses this endpoint yet so I am not sure what are the applications yet. Also, performing a linear scan with signature verification in an unbounded list scares me, so I'd rather have the caller making these checks.

File: beacon-chain/operations/blstoexec/pool.go (hunk 2 above)
Comment (potuz) [reply beacon-chain/operations/blstoexec/pool.go:81]: MarkIncluded needs a lock, so we can't defer.

File: beacon-chain/operations/blstoexec/pool.go (hunk 1 above)
Comment (terencechain) [reply beacon-chain/operations/blstoexec/pool.go:21]: That's fair. I don't know whether beacon API explicitly calls this out. We can add this behavior to the commentary  if it doesn't

**Review by terencechain — APPROVED** [review]
lgtm

**Comment by terencechain** [comment]
Please don't merge it. A few more changes are coming

File: beacon-chain/operations/blstoexec/pool.go [hunk 5]
```diff
@@ -58,25 +63,70 @@ func (p *Pool) PendingBLSToExecChanges() ([]*ethpb.SignedBLSToExecutionChange, e
...
+	cSet, err := blocks.BLSChangesSignatureBatch(st, result)
+	if err != nil {
+		logrus.WithError(err).Warning("could not get BLSToExecutionChanges signatures")
+	} else {
+		ok, err := cSet.Verify()
+		if err != nil {
+			logrus.WithError(err).Warning("could not batch verify BLSToExecutionChanges signatures")
+		} else if ok {
+			return result, nil
+
```
Comment (rkapka) [inline beacon-chain/operations/blstoexec/pool.go:109]:
```suggestion
```

File: beacon-chain/operations/blstoexec/pool.go [hunk 6]
```diff
@@ -58,25 +63,70 @@ func (p *Pool) PendingBLSToExecChanges() ([]*ethpb.SignedBLSToExecutionChange, e
...
+			result = append(result, change)
+		}
 		node, err = node.Next()
 		if err != nil {
+			p.lock.RUnlock()
 			return nil, err
 		}
 	}
-	return result, nil
+	p.lock.RUnlock()
```
[the remark this answers is not in the document]
Comment (potuz) [reply beacon-chain/operations/blstoexec/pool.go:95]: discussed offline

**Review by terencechain — APPROVED** [review]

**Review by rkapka — APPROVED** [review]

**Pull request merged**